README.md
# Button Platform for Homebridge

[![verified-by-homebridge](https://badgen.net/badge/homebridge/verified/purple)](https://github.com/homebridge/homebridge/wiki/Verified-Plugins)

This Homebridge platform plugin creates virtual buttons that can be triggered
using HTTP requests. It was originally created to provide HomeKit support for
[Flic smart buttons](https://flic.io/) before the Flic Hub LR got native HomeKit
support but can be leveraged by anything that can send HTTP requests.

> **Flic Hub LR now supports HomeKit**: Flic released a firmware update for the
Flic Hub LR on 2 December 2020 which included native HomeKit support.
Visit <https://flic.io/homekit> for instructions on how to configure native
HomeKit access.

**This plugin requires a working home hub.**

Before you install, ensure you have [setup a HomePod, HomePod mini, Apple TV or
iPad as a home hub](https://support.apple.com/en-au/HT207057) for HomeKit. A
home hub is required to run automations triggered by the button events generated
by this plugin.

## Install

The simplest method to install and configure this plugin is via
[`homebridge-config-ui-x`](https://www.npmjs.com/package/homebridge-config-ui-x).

To install manually, run the following command in your Homebridge directory.
Depending on how you installed Homebridge, you may need to add the `-g` and/or
the `--unsafe-perms` parameters:

```shell
npm install [-g|--unsafe-perms] homebridge-button-platform
```

The `-g` option will install the plugin globally and the `--unsafe-perms` option
is needed for some platforms to install successfully.

## Configuration

The plugin can be configured via the [`homebridge-config-ui-x`](https://www.npmjs.com/package/homebridge-config-ui-x)
admin interface.

To configure the plugin manually, add the following configuration to the
`platforms` block of your Homebridge `config.json` file:

```json
"platforms": [
    {
        "platform": "button-platform",
        "port": 3001,
        "buttons": [
            "My Button",
            "Your Button",
            "Their Button"
        ]
    }
  ]
```

You can add as many buttons to the array as you need. Each button will get its
own URI on which the server will listen for events. You can determine the URI
for each button by checking the Homebridge logs for
`The Event URI for <button name> is: /button-button-name` strings.

> **Note:** renaming a button in `config.json` is the same as deleting the old
button and adding a new one, i.e. any configuration will be lost.

## Sending events

To trigger a button event, send an HTTP `GET` request with a specific query
string value or an HTTP `POST` request to your Homebridge hostname or IP address
and the port specified in the configuration of the platform, plus the URI for the
button.

### Using a `GET` request

The simplest method of triggering an event is to use a simple `GET` request with
the query string parameter of `event` set to one of the valid event types:

* `click` or `single-press`
* `double-click` or `double-press`
* `hold` or `long-press`

For example, to send a single click event to a button using `curl`:

```shell
$ curl http://<homebridge>:<port>/uri?event=click
Success
```

You could also use this URL with a normal web browser.

### Example Flic button configuration

Here’s a screenshot of what an Internet Request action should look like in the
Flic app if your Homebridge server is `homebridge.local` and the plugin
was listening on port 3001:

![flic-config](docs/flic-config.png)

The values are as follows:

| Field | Value |
|:------|:------|
| Hub Action | **Internet Request** |
| URL | `http://homebridge.local:3001/button-name?event=click` |
| Type | **GET** |

### Battery level

The plugin from v1.5 onwards automatically creates a
battery service when it detects it's being triggered by a Flic button and will
update the battery level with the current value from the Flic whenever the
button is pushed.

### Example Stream Deck configuration

You can use the "Website" Stream Deck action to trigger an event. In the
Stream Deck configuration, drag the "Website" action to a key and configure
the URL to be `http://homebridge.local:3001/button-name?event=click` which uses the
same parameters as the `GET` request above:

![streamdeck-key](docs/streamdeck-config.png)

> **Top tip:** If you enable the `GET request in background` option, a browser will
> not open when you hit the key.

## Using a `POST` request

If you use an HTTP `POST` request you must set the Content-Type header to either
`application/json` or `application/x-www-form-urlencoded`. The body of the request
needs a field named `event` with a value of one of the event types defined above.

For example, to send a double press event to a button using `curl` with a JSON
payload:

```shell
$ curl -X POST \
  -H 'Content-Type: application/json' \
  -d '{"event": "double-click"}' \
  http://<homebridge>:<port>/<uri>
```

You can also `POST` with an `application/x-www-form-urlencoded` payload:

```shell
$ curl -X POST \
  -H 'Content-Type: application/x-www-form-urlencoded' \
  -d 'event=double-click' \
  http://<homebridge>:<port>/<uri>
```

This is useful when used in with a larger form-based application.

## Troubleshooting

Check the Homebridge logs for any warnings as the plugin will log any attempts
to trigger an invalid event or any invalid URIs.


CHANGELOG.md
# Change Log

All notable changes to this project will be documented in this file.

## 1.5.1 (2021-12-09)

* The battery is now set to 100% whenever Homebridge starts and will only
  change when it gets a new value via a header (Fixes #88).

## 1.5.0 (2021-08-31)

* Significant refactoring to make the plugin cleaner and easier to maintain
* Updated lots of dependencies

## 1.4.3 (2021-05-23)

* Updated [`homebridge-lib`](https://github.com/ebaauw/homebridge-lib) to v5.1.3
  which requires:
  * homebridge: v1.3.4
  * Node.js: 14.17.0
* Code cleanup and review

## 1.4.2 (2021-05-21)

* Fix #46: `POST` requests were not calling the right function
* Updated `express-validator` dependency to v6.11.1

## 1.4.1 (2021-04-19)

Buttons have batteries! The virtual button will now reflect the current
battery level if triggered by a Flic 2 button. A battery level of less than
15% is flagged as low.

The patch version bump to 1.4.1 was caused by me breaking the publish automation
and having to do it twice.

## 1.3.0 (2021-02-26)

Bump dependencies as HAP-NodeJS is not setting eventOnlyCharacteristic properly
and homebridge-lib 5.0.9 includes a workaround for this.

## 1.2.0 (2020-11-10)

Bump dependencies and minimum requirements:

* [`homebridge-lib`](https://github.com/ebaauw/homebridge-lib) to v4.8.0 which requires:
  * Node.js v14.15.0 LTS
  * Homebridge 1.1.0

## 1.1.5 (2020-10-10)

Bump dependencies and minimum requirements:

* [`homebridge-lib`](https://github.com/ebaauw/homebridge-lib) to v4.7.16 which requires:
  * Node.js v12.19.0 LTS
  * Homebridge 1.1.0

## 1.1.4 (2020-07-12)

Bump dependencies:

* [`homebridge-lib`](https://github.com/ebaauw/homebridge-lib) to v4.7.12
* Node.js v12.18.2 LTS;
* bonjour-hap v3.5.11.

## 1.1.3 (2020-06-01)

* Disable heartbeat for each button.
* Remove check for configuration as Homebridge handles this automatically.
* Update `homebridge-lib` dependency to `4.7.7`.
* Updated minimum Homebridge version to `1.1.0`.
* Updated minimum Node.js version to `12.17.0 LTS`.

## 1.1.2 (2020-05-24)

* Remove stale buttons automatically
* Update README.md to indicate what happens on button rename

## 1.1.0 (2020-05-20)

* Enable `GET` requests to trigger button events
* Update `homebridge-lib` dependency to `4.7.5`
* Update `express-validator` dependency to `6.5.0`

## 1.0.5 (2020-05-01)

* Updated `homebridge-lib` dependency to `4.6.1`.

## 1.0.4 (2020-03-15)

* Removed `semistandard` as a development dependency.

## 1.0.3 (2020-03-15)

* Updated `homebridge-lib` dependency to `4.5.5`.

## 1.0.2 (2020-03-15)

* Removed the custom `UUIDGen` function as `homebridge-lib` handles this
  internally.

## 1.0.1 (2020-03-15)

* Initial release to NPM as `homebridge-button-platform`
* Enable config via `homebridge-config-ui-x`

## 1.0.0

* Initial release


index.js
/*
 *  Copyright 2021 Avi Miller
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

"use strict";

const ButtonPlatform = require("./lib/ButtonPlatform");
const PackageJson = require("./package.json");

module.exports = function (homebridge) {
  ButtonPlatform.loadPlatform(
    homebridge,
    PackageJson,
    "button-platform",
    ButtonPlatform
  );
};


lib/FlicPlatform.js
'use strict';

const express = require('express');
const { check, validationResult } = require('express-validator');
const events = require('events');
const HomebridgeLib = require('homebridge-lib');
const FlicAccessory = require('./FlicAccessory');

class FlicPlatform extends HomebridgeLib.Platform {
  constructor (log, configJson, homebridge) {
    super(log, configJson, homebridge);
    if (configJson == null) {
      return;
    }
    this.on('accessoryRestored', this.accessoryRestored);
    this.once('heartbeat', this.init);

    this.config = {
      name: 'Flic',
      port: 3001,
      buttons: []
    };

    const optionParser = new HomebridgeLib.OptionParser(this.config, true);
    optionParser.stringKey('name');
    optionParser.intKey('port', 1025, 65535);
    optionParser.listKey('buttons');
    optionParser.on('usageError', (message) => {
      this.warn('Configuration issue: %s', message);
    });

    try {
      optionParser.parse(configJson);
      if (this.config.buttons.length === 0) {
        this.warn('Configuration issue: no buttons configured.');
      }

      this.app = express();
      this.app.listen(this.config.port, () => this.log('Listening on port %s for inbound Flic button push event notifications', this.config.port));
      this.app.use(express.json());

      this.flicAccessories = {};
    } catch (error) {
      this.fatal(error);
    }
  }

  async init (beat) {
    const jobs = [];
    for (const button of this.config.buttons) {
      if (this.flicAccessories[button] == null) {
        const flicAccessory = new FlicAccessory(this, { button: button });
        jobs.push(events.once(flicAccessory, 'initialised'));
        this.flicAccessories[button] = flicAccessory;
        this.setupRoute(this.flicAccessories[button]);
      } else {
        this.flicAccessories[button].setAlive();
      }
    }
    for (const job of jobs) {
      await job;
    }
    this.debug('initialised');
    this.emit('initialised');
  }

  accessoryRestored (className, version, id, name, context) {
    if (className !== 'FlicAccessory') {
      this.warn(
        'removing cached %s accessory %s',
        className, context.button
      );
      return;
    }
    const flicAccessory = new FlicAccessory(this, context);
    this.flicAccessories[context.button] = flicAccessory;
    this.setupRoute(flicAccessory);
  }

  setupRoute (accessory) {
    const uri = '/flic-' + accessory.name.toLowerCase().replace(/[^a-z0-9]/gi, '-');
    this.debug('creating route for %s via %s', accessory.name, uri);

    this.app.post(uri, [
      check('event').isIn(['click', 'double-click', 'hold'])
    ], (req, res) => {
      const errors = validationResult(req);
      if (!errors.isEmpty()) {
        return res.status(422).json({ errors: errors.array() });
      } else {
        var event = req.body.event;
        res.status(200).send('Success');
        this.log('Received POST request on %s to trigger a [%s] event for %s', uri, event, accessory.name)
      }
    });
  }
}

module.exports = FlicPlatform;


lib/ButtonService.js
/*
 *  Copyright 2021 Avi Miller
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

'use strict';

const HomebridgeLib = require('homebridge-lib');

class ButtonService extends HomebridgeLib.ServiceDelegate {
  static get StatelessProgrammableSwitch() {
    return StatelessProgrammableSwitch;
  }
}

class StatelessProgrammableSwitch extends ButtonService {
  constructor(buttonAccessory, params = {}) {
    params.name = buttonAccessory.name + ' Switch';
    params.Service = buttonAccessory.Services.hap.StatelessProgrammableSwitch;
    super(buttonAccessory, params);
    this.addCharacteristicDelegate({
      key: 'programmableSwitchEvent',
      Characteristic: this.Characteristics.hap.ProgrammableSwitchEvent
    });
  }

  triggerEvent(event) {
    this.values.programmableSwitchEvent = event;
  }
}

module.exports = ButtonService;


lib/FlicAccessory.js
'use strict';

const crypto = require('crypto');
const HomebridgeLib = require('homebridge-lib');
//const FlicService = require('./FlicService');

function generateUUID (data) {
  // https://github.com/KhaosT/HAP-NodeJS/blob/master/src/lib/util/uuid.ts#L7
  const sha1sum = crypto.createHash('sha1');
  sha1sum.update(data);
  const s = sha1sum.digest('hex');
  let i = -1;
  return 'xxxxxxxx-xxxx-4xxx-yxxx-xxxxxxxxxxxx'.replace(/[xy]/g, (c) => {
    i += 1;
    switch (c) {
      case 'y':
        return ((parseInt('0x' + s[i], 16) & 0x3) | 0x8).toString(16);
      case 'x':
      default:
        return s[i];
    }
  });
}

class FlicAccessory extends HomebridgeLib.AccessoryDelegate {
  constructor (platform, context) {
    const params = {
      name: context.button,
      id: generateUUID(context.button.toLowerCase().replace(/[^a-z0-9]/gi, '-')),
      manufacturer: 'Shortcut Labs',
      model: 'Flic 2',
      category: platform.Accessory.Categories.Switch
    };
    super(platform, params);

    this.context.button = context.button;

    setImmediate(() => {
      this.emit('initialised');
    });
  }
}

module.exports = FlicAccessory;


lib/BatteryService.js
/*
 *  Copyright 2021 Avi Miller
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

'use strict';

const HomebridgeLib = require('homebridge-lib');

class BatteryService extends HomebridgeLib.ServiceDelegate {
  constructor(buttonAccessory, params) {
    super(buttonAccessory, params);
    this.buttonAccessory = buttonAccessory;
  }
  static get Battery() {
    return Battery;
  }
}

class Battery extends BatteryService {
  constructor(buttonAccessory, params = {}) {
    params.name = buttonAccessory.name + ' Battery';
    params.Service = buttonAccessory.Services.hap.Battery;
    super(buttonAccessory, params);

    this.addCharacteristicDelegate({
      key: 'batteryLevel',
      Characteristic: this.Characteristics.hap.BatteryLevel
    });
    this.addCharacteristicDelegate({
      key: 'statusLowBattery',
      Characteristic: this.Characteristics.hap.StatusLowBattery
    });
    this.addCharacteristicDelegate({
      key: 'chargingState',
      Characteristic: this.Characteristics.hap.ChargingState,
      silent: true,
      value: this.Characteristics.hap.ChargingState.NOT_CHARGEABLE
    });

    this.updateBattery(100);
  }

  updateBattery(batteryLevel) {
    this.values.batteryLevel = batteryLevel;

    // Set the low battery status indicator if there is less than 15% battery
    this.values.statusLowBattery =
      batteryLevel < 15
        ? this.Characteristics.hap.StatusLowBattery.BATTERY_LEVEL_LOW
        : this.Characteristics.hap.StatusLowBattery.BATTERY_LEVEL_NORMAL;
  }
}

module.exports = BatteryService;


lib/ButtonAccessory.js
/*
 *  Copyright 2021 Avi Miller
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

'use strict';

const HomebridgeLib = require('homebridge-lib');
const ButtonService = require('./ButtonService');
const BatteryService = require('./BatteryService');

class ButtonAccessory extends HomebridgeLib.AccessoryDelegate {
  constructor(platform, context) {
    const params = {
      name: context.button,
      id: context.button.toLowerCase().replace(/[^a-z0-9]/giu, '-'),
      manufacturer: 'Homebridge',
      model: 'Virtual Button',
      category: platform.Accessory.Categories.Switch
    };
    super(platform, params);

    this.context.button = context.button;
    this.batteryService = new BatteryService.Battery(this);
    this.switchService = new ButtonService.StatelessProgrammableSwitch(this);

    this.heartbeatEnabled = false;

    setImmediate(() => {
      this.emit('initialised');
    });
  }
}

module.exports = ButtonAccessory;


lib/ButtonPlatform.js
/*
 *  Copyright 2021 Avi Miller
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

'use strict';

const express = require('express');
const helmet = require('helmet');

const {check, header, oneOf, validationResult} = require('express-validator');
const events = require('events');
const HomebridgeLib = require('homebridge-lib');
const ButtonAccessory = require('./ButtonAccessory');

class ButtonPlatform extends HomebridgeLib.Platform {
  constructor(log, configJson, homebridge) {
    super(log, configJson, homebridge);

    this.once('heartbeat', this.init);

    this.config = {
      name: 'Buttons',
      port: 3001,
      buttons: []
    };

    const optionParser = new HomebridgeLib.OptionParser(this.config, true);
    optionParser.stringKey('name');
    optionParser.intKey('port', 1025, 65535);
    optionParser.listKey('buttons');
    optionParser.on('usageError', message => {
      this.warn('Configuration issue: %s', message);
    });

    try {
      optionParser.parse(configJson);
      if (this.config.buttons.length === 0) {
        this.warn('Configuration issue: no buttons configured.');
      }

      this.app = express();
      this.app.listen(this.config.port, () =>
        this.log('Listening on port %s for inbound button push event notifications', this.config.port)
      );
      this.app.use(helmet());
      this.app.use(express.json());
      this.app.use(express.urlencoded({extended: false}));

      this.buttonAccessories = {};
    } catch (error) {
      this.fatal(error);
    }
  }

  async init() {
    const jobs = [];
    for (const button of this.config.buttons) {
      if (this.buttonAccessories[button] === undefined) {
        const buttonAccessory = new ButtonAccessory(this, {button: button});
        jobs.push(events.once(buttonAccessory, 'initialised'));
        this.setupRoute(buttonAccessory);
        this.buttonAccessories[button] = buttonAccessory;
      }
    }
    for (const job of jobs) {
      await job;
    }

    // Express handler for invalid paths
    this.app.use(
      function (req, res) {
        res.status(404).send('Button not found.');
        this.warn('Received event for unconfigured Button path: %s', req.originalUrl);
      }.bind(this)
    );

    // Express handler for server-side errors
    this.app.use(
      function (err, req, res) {
        res.status(500).send('Server error.');
        this.fatal(err.stack);
      }.bind(this)
    );

    this.debug('initialised');
    this.emit('initialised');
  }

  setupRoute(accessory) {
    const uri = '/button-' + accessory.name.toLowerCase().replace(/[^a-z0-9]/giu, '-');
    this.log('The Event URI for %s is: %s', accessory.name, uri);

    this.app.all(
      uri,
      [
        check('event').isIn(['click', 'double-click', 'hold', 'single-press', 'double-press', 'long-press']),
        oneOf(
          [header('button-battery-level').not().exists(), header('button-battery-level').isInt({min: 0, max: 100})],
          'button-battery-level must be an integer between 0 and 100 which represents the percentage battery level remaining.'
        )
      ],
      (req, res) => {
        const errors = validationResult(req);
        if (!errors.isEmpty()) {
          return res.status(422).json({errors: errors.array()});
        } else {
          const event = req.query.event || req.body.event;

          switch (event) {
            case 'click':
            case 'single-press':
              accessory.switchService.triggerEvent(0);
              break;
            case 'double-click':
            case 'double-press':
              accessory.switchService.triggerEvent(1);
              break;
            case 'hold':
            case 'long-press':
              accessory.switchService.triggerEvent(2);
              break;
          }

          if (req.headers['button-battery-level']) {
            accessory.batteryService.updateBattery(req.headers['button-battery-level']);
          }

          res.status(200).send('OK');
        }
      }
    );
  }
}

module.exports = ButtonPlatform;
